### DISCUSSION TITLE
more options for sea-orm-cli

### DISCUSSION THREAD

**Discussion by vincascm** [SeaQL/sea-orm#451] [Ideas]
i use `sea-orm-cli` to generate `model`，and the work flow like update database schema, sea-orm-cli generate, web api `mod` use it.
with this work flow, generated code is immutable，but may be add more options to fit it.
my English is too poor, forgive me for using Chinese.
发现的问题有几个:
1. sea-orm-cli generate的 output-dir 是 `src/model`，在这个目录里，我会有一个`impls`目录，用来放`Model`额外的implement，所以需要model中mod.rs的生成是可选的
2. prelude.rs的生成是否可选，我觉得并不需要prelude，都是直接通过mod名称引用的
3. PartialEq的derive是否可选，我需要自定义排序
4. 需要将某些字段的类型指定为Json，比如MySQL的text类型

**Comment by billy1624** [comment]
Hey @vincascm, welcome!

I'm translating your comment with Google Translate, please do let me know if the translation is inaccurate.

> i use sea-orm-cli to generate model, and the work flow like update database schema, sea-orm-cli generate, web api mod use it.
> with this work flow, generated code is immutable, but may be add more options to fit it.
> my English is too poor, forgive me for using Chinese.
> There are several problems found:
> 
> 1. The output-dir of sea-orm-cli generate is `src/model`. In this directory, I will have an `impls` directory to store additional implements of the `Model`, so the generation of mod.rs in the model is optional.
> 2. Whether the generation of prelude.rs is optional, I don’t think prelude is needed, all are directly referenced by mod name
> 3. Whether the derive of PartialEq is optional, I need to customize the sort
> 4. Need to specify the type of some fields as Json, such as the text type of MySQL

Reply:

> 1. The output-dir of sea-orm-cli generate is `src/model`. In this directory, I will have an `impls` directory to store additional implements of the `Model`, so the generation of mod.rs in the model is optional.

Hmmm... okay so you want to organize the code in a different way. Could you clarify what will be stored in `impls` directory?

> 2. Whether the generation of prelude.rs is optional, I don’t think prelude is needed, all are directly referenced by mod name

Yeah, this could be optional but this is just one file, so I guess user can delete it manually if they don't want it. We could add an option to skip the generation of it if you really want to.

> 3. Whether the derive of PartialEq is optional, I need to customize the sort

We would add an option for it as well.

> 4. Need to specify the type of some fields as Json, such as the text type of MySQL

This has to be carefully consider. The part of code responsible of it:

https://github.com/SeaQL/sea-orm/blob/cc4904993a190e827f3031210bd965e6f46c9165/sea-orm-codegen/src/entity/column.rs#L25-L55

**Comment by vincascm** [comment]
@billy1624  thanks your reply,
1. add some `mod` in `src/model`,  it's just a pattern such as impl extra methods for model, to express more precisely business logic on model, and this layer encapsulates closest to CRUD of model, so i think it should be in `src/model`
2.  the Json here refers to a converted type, which in the database may be TEXT, BLOB, or other types that can be converted to Json. If there is no additional information, these types are converted to directly corresponding types, such as String or `Vec <u8>`
